### PR TITLE
#165305223 Users should be able to Login to Authors Haven

### PR DESCRIPTION
**Pull request by Benkimeric**
#### What does this PR do?
- Implements user functionality to allow a user to login

#### Description of Task to be completed?
- Add functionality to log in a user
- Add functionality to log out a user
- Add tests for use login functionality

#### How should this be manually tested?

- [x] **Clone the repository**

        $ git clone https://github.com/andela/ah-the-jedi-frontend.git

- [x] **Switch to the ah-the-jedi-frontend directory**

       $ cd ah-the-jedi-frontend/

- [x] **Switch to testing branch**

       $ ft-implement-user-login-165305223

 
- [x] **Install the project dependencies**

      $  npm install

- [x] **Running the project**

       $ npm run start:dev

- [x] **Navigate to the login page**

       $ <your-localhost>/login
     

#### What are the relevant pivotal tracker stories?
[#165305223](https://www.pivotaltracker.com/story/show/165305223)
#### Any background context you want to add?
None
#### Successful Login
successful login

#### Unsuccessful on login
unsuccessful login

### CODE REVIEW

**Review by ElishaMisoi — CHANGES_REQUESTED** [review]
Good implementation @Benkimeric 👍🏽. You should however squash your commits.

**Review by AlvinMugambi — APPROVED** [review]
Great implementation. Good work on styling the login box.

**Review by ElishaMisoi — APPROVED** [review]

**Review by katherine95 — APPROVED** [review]
LGTM

**Review by Siffersari — APPROVED** [review]
Good job. I love the use of toasts to display the error messages and the use of loader to show that the request is underway. Great implementation for deactivated login button for validation. Really intuitive.

File: src/assets/styles/login.scss [hunk 1]
```diff
@@ -1,4 +1,77 @@
...
+      width: 48%;
+    }
+    &-text {
+      background: transparent;
+      border-radius: 50%;
+      margin: 0 5px 0 5px;
+      border: 2px solid #cecece;
+      padding: 1%;
+    }
+    margin-bottom: 4%;
```
Comment (AnguleMathias) [inline src/assets/styles/login.scss:46]: Move line `47-48` to be on line `39`

File: src/assets/styles/login.scss [hunk 2]
```diff
@@ -1,4 +1,77 @@
...
+    display: none;
+  }
+
+  .show {
+    display: block;
+  }
+  @media (min-width: 992px) {
+    .signin-form {
+      display: inline-block;
+      -moz-box-shadow: 1px 2px 4px rgba(0, 0, 0, 0.5);
```
Comment (AnguleMathias) [inline src/assets/styles/login.scss]: Store the color spaces in a `variables.scss` file and import the variable name

File: src/components/App.js [hunk 3]
```diff
@@ -1,12 +1,15 @@
 import React, { Component, Fragment } from 'react';
 import { BrowserRouter as Router, Route, Switch } from 'react-router-dom';
+import { ToastContainer } from 'react-toastify';
 import Header from './layout/Header';
 import routes from '../routes';
+import 'react-toastify/dist/ReactToastify.css';
```
Comment (AnguleMathias) [inline src/components/App.js]: move this to line 4

File: src/components/auth/LoginForm.js [hunk 4]
```diff
@@ -0,0 +1,80 @@
+import React, { Component } from 'react';
+import '../../assets/styles/login.scss';
```
Comment (AnguleMathias) [inline src/components/auth/LoginForm.js]: move down to line 4

File: src/components/layout/LoggedInLinks.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
+import React from 'react';
+import { NavLink, withRouter } from 'react-router-dom';
+import { connect } from 'react-redux';
+import { logoutUser } from '../../redux/actions/loginActions';
+
+class LoggedInLinks extends React.Component {
```
Comment (AnguleMathias) [inline src/components/layout/LoggedInLinks.js:10]: Add documentation eg
```
/*
* LoggedInLinks Component
*
*@return {jsx}
*/

```

File: tests/App.test.jsx [hunk 6]
```diff
@@ -1,11 +1,21 @@
 import React from 'react';
 import ReactDOM from 'react-dom';
 import App from '../src/components/App';
+import { BrowserRouter } from 'react-router-dom';
+import { Provider } from 'react-redux';
+import store from '../src/redux/store/';
 
 describe('PASSES', () => {
```
Comment (AnguleMathias) [inline tests/App.test.jsx]: Have the describe as:
`Unit tests for App Component`

File: tests/LoggedInLinks.test.jsx [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    );
+    return component;
+  };
+
+  const findByAttribute = (component, attr) => {
+    const wrapper = component.find(`[data-set='${attr}']`);
+    return wrapper;
+  };
+
+  describe('Test for the Header component <LoggedInLinks />', () => {
```
Comment (AnguleMathias) [inline tests/LoggedInLinks.test.jsx]: have it as:

`Unit tests for Header Component`

File: tests/LoggedInLinks.test.jsx [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    let component;
+    beforeEach(() => {
+      component = setUp();
+    });
+    it('Should render component without errors', () => {
```
Comment (AnguleMathias) [inline tests/LoggedInLinks.test.jsx]: it should render nav-bar-test

**Review by AnguleMathias — CHANGES_REQUESTED** [review]
Great job @Benkimeric on implementing the feature, kindly check on the comments

**Review by jkamz — APPROVED** [review]

**Review by ElishaMisoi — APPROVED** [review]

**Pull request merged**